python/aws_lambda_powertools/middleware_factory/factory.py

```diff
@@ -0,0 +1,143 @@
+            any_code_to_execute_before_lambda_handler()
+            response = handler(event, context)
+            any_code_to_execute_after_lambda_handler()
+            print(f"Lambda handler response: {response}")
+
+        @log_response
+        def lambda_handler(event, context):
+            return True
+
+    **Create a middleware with params**
+
+        from aws_lambda_powertools.middleware_factory import lambda_handler_decorator
+
+        @lambda_handler_decorator
+        def obfuscate_sensitive_data(handler, event, context, fields=None):
+            # Obfuscate email before calling Lambda handler
+            if fields:
+                for field in fields:
+                    field = event.get(field, "")
+                    event[field] = obfuscate_pii(field)
+
+            response = handler(event, context)
+            print(f"Lambda handler response: {response}")
+
+        @obfuscate_sensitive_data(fields=["email"])
+        def lambda_handler(event, context):
+            return True
+
+    **Trace execution of custom middleware**
+
+        from aws_lambda_powertools.middleware_factory import lambda_handler_decorator
+
+        @lambda_handler_decorator(trace_execution=True)
+        def log_response(handler, event, context):
+            ...
+
+        @log_response
+        def lambda_handler(event, context):
+            return True
+
+    Limitations
+    -----------
+    * Async middlewares not supported
+    * Classes, class methods middlewares not supported
+
+    Raises
+    ------
+    TypeError
+        When middleware receives non keyword=arguments
+    """
+
+    if decorator is None:
+        return functools.partial(lambda_handler_decorator, trace_execution=trace_execution)
+
+    trace_execution = trace_execution or strtobool(str(os.getenv("POWERTOOLS_TRACE_MIDDLEWARES", False)))
+
+    @functools.wraps(decorator)
+    def final_decorator(func: Callable = None, **kwargs):
+        # If called with kwargs return new func with kwargs
+        if func is None:
+            return functools.partial(final_decorator, **kwargs)
+
+        if not inspect.isfunction(func):
+            raise TypeError(
+                f"Only keyword arguments is supported for middlewares: {decorator.__qualname__} received {func}"
+            )
+
+        @functools.wraps(func)
+        def wrapper(event, context):
+            try:
+                middleware = functools.partial(decorator, func, event, context, **kwargs)
+                if trace_execution:
+                    with _trace_middleware(middleware=decorator):
+                        response = middleware()
+                else:
+                    response = middleware()
+                return response
+            except Exception as err:
+                logger.error(f"Caught exception in {decorator.__qualname__}")
+                raise err
+
+        return wrapper
+
+    return final_decorator
+
+
+@contextmanager
+def _trace_middleware(middleware):
```

Comment:
Are there plans to use this beyond the `lambda_handler_decorator` function? Otherwise, shouldn't we just remove this and move the code into the `lambda_handler_decorator` function?

Reply:
Good catch! I actually had when i first wrote! I had created another decorator to be used in classes to reduce LOC but since abandoned that.

Will address this shortly.